test/OpenFeature.Tests/OpenFeatureEventTests.cs

```diff
@@ -0,0 +1,413 @@
+
+                eventHandler.Invoke(Arg.Any<ProviderEventPayload>());
+
+                Api.Instance.AddHandler(ProviderEventTypes.ProviderReady, failingEventHandler);
+                Api.Instance.AddHandler(ProviderEventTypes.ProviderReady, eventHandler);
+
+                var testProvider = new TestProvider(fixture.Create<string>());
+                await Api.Instance.SetProvider(testProvider);
+
+                failingEventHandler.Received().Invoke(Arg.Is<ProviderEventPayload>(payload => payload.ProviderName == testProvider.GetMetadata().Name));
+                eventHandler.Received().Invoke(Arg.Is<ProviderEventPayload>(payload => payload.ProviderName == testProvider.GetMetadata().Name));
+            }
+            catch (Exception e)
+            {
+                // make sure the exception is passed on correctly
+                Assert.NotNull(e);
+            }
+        }
+
+        [Fact]
+        [Specification("5.1.2", "When a `provider` signals the occurrence of a particular `event`, the associated `client` and `API` event handlers MUST run.")]
+        [Specification("5.2.1", "The `client` MUST provide a function for associating `handler functions` with a particular `provider event type`.")]
+        [Specification("5.2.3", "The `event details` MUST contain the `provider name` associated with the event.")]
+        [Specification("5.2.4", "The `handler function` MUST accept a `event details` parameter.")]
+        public async Task Client_Level_Event_Handlers_Should_Be_Registered()
+        {
+            var fixture = new Fixture();
+            var eventHandler = Substitute.For<EventHandlerDelegate>();
+
+            eventHandler.Invoke(Arg.Any<ProviderEventPayload>());
+
+            var myClient = Api.Instance.GetClient(fixture.Create<string>());
+
+            var testProvider = new TestProvider();
+            await Api.Instance.SetProvider(myClient.GetMetadata().Name, testProvider);
+
+            myClient.AddHandler(ProviderEventTypes.ProviderReady, eventHandler);
+
+            eventHandler.Received().Invoke(Arg.Is<ProviderEventPayload>(payload => payload.ProviderName == testProvider.GetMetadata().Name));
+        }
+
+        [Fact]
+        [Specification("5.1.2", "When a `provider` signals the occurrence of a particular `event`, the associated `client` and `API` event handlers MUST run.")]
+        [Specification("5.2.1", "The `client` MUST provide a function for associating `handler functions` with a particular `provider event type`.")]
+        [Specification("5.2.3", "The `event details` MUST contain the `provider name` associated with the event.")]
+        [Specification("5.2.4", "The `handler function` MUST accept a `event details` parameter.")]
+        [Specification("5.2.5", "If a `handler function` terminates abnormally, other `handler` functions MUST run.")]
+        public async Task Client_Level_Event_Handlers_Should_Be_Executed_When_Other_Handler_Fails()
+        {
+            try
```

Comment:
I'm confused by this try/catch. It seems unnecessary, and we wouldn't want handlers to throw (indeed it seems like you catch and swallow these errors as I'd expect)

Reply:
good catch (... ;-) ) That was a left over from trying out different approaches on how to best handle exceptions thrown by the event handlers, but this can be removed now